File: server1/server1.go
package main

import (
	"fmt"
	"io/ioutil"

	"github.com/gin-gonic/gin"
	graphql "github.com/graph-gophers/graphql-go"
	"github.com/graph-gophers/graphql-go/relay"
)

var fakeDatabase = map[string]string{}

type root struct{}

func (r *root) GetMessage(args struct{ Key string }) *string {
	s := fakeDatabase[args.Key]
	return &s
}

func (r *root) SetMessage(args struct{ Key, Message string }) *string {
	fakeDatabase[args.Key] = args.Message
	s := fakeDatabase[args.Key]
	return &s
}

// Construct a schema, using GraphQL schema language
var schemaBytes, _ = ioutil.ReadFile("server1.graphql")
var schema = graphql.MustParseSchema(string(schemaBytes), &root{}, graphql.UseStringDescriptions())

// GIN handler
func ginGraphQLHandler() gin.HandlerFunc {
	r := &relay.Handler{Schema: schema}
	return func(c *gin.Context) {
		r.ServeHTTP(c.Writer, c.Request)
	}
}

func main() {
	fmt.Println("Running a GraphQL API server at http://localhost:4000/graphql")
	r := gin.New()
	r.POST("/graphql", ginGraphQLHandler())
	r.Run(":4000")
}

File: server1/server1.js
// Running GraphQL API server
var fs = require('fs')
var express = require('express')
var { graphqlHTTP } = require('express-graphql')
var { buildSchema } = require('graphql')
var cors = require('cors')

var fakeDatabase = { k0: 'Hello World!' }

// The root provides the top-level API endpoints
var root = {
    getMessage: ({ key }) => {
        return fakeDatabase[key]
    },
    setMessage: ({ key, message }) => {        
        return fakeDatabase[key] = message
    }
}

// Construct a schema, using GraphQL schema language
var schema = buildSchema(fs.readFileSync('server1.graphql', 'utf8'))

var app = express()
app.use(cors())
app.use( '/graphql', graphqlHTTP({ 
        schema: schema, 
        rootValue: root, 
        graphiql: true, 
    })
)
console.log('Running a GraphQL API server at http://localhost:4000/graphql')
app.listen(4000)
